core/src/main/java/org/jdbi/v3/core/result/ResultIterable.java

```diff
@@ -303,4 +315,103 @@ default <U> U reduce(U identity, BiFunction<U, T, U> accumulator) {
                 });
         }
     }
+
+    /**
+     * Filters the {@link ResultIterable} by applying the specified predicate.<br>
+     * This method has the look and feel of {@link Stream#filter(Predicate)} without making use of streams.
+     *
+     * @param predicate a non-interfering, non-null predicate to apply to each element
+     *                  to determine whether it should be included in the result
+     * @return the new result iterable
+     *
+     * @since 3.30.1
+     */
+    default ResultIterable<T> filter(Predicate<? super T> predicate) {
+        Objects.requireNonNull(predicate, "Filter required");
+        return () -> new ResultIteratorDelegate<T, T>(iterator()) {
+            /** The next result (initially {@code null}). */
+            private T next;
+
+            @Override
+            public boolean hasNext() {
+                return findNext();
+            }
+
+            /**
+             * Returns the next result in the resultset.<br>
+             * {@link #hasNext()} and this method synchronize on {@code this} internally
+             * as both forward the resultset and read/write member variable {@link #next}.
+             *
+             * @return the next element in the resultset that passes the filter
+             * @throws NoSuchElementException if the resultset has no more suitable elements
+             */
+            @Override
+            public T next() {
+                synchronized (this) {
+                    if (next == null && !findNext()) {
+                        throw new NoSuchElementException("No more filtered results");
+                    }
+                    T n = next;
+                    next = null;
+                    return n;
+                }
+            }
+
+            /**
+             * Forwards the resultset to the next result that passes the filter (i.e. {@link Predicate} tests {@code true}).<br>
+             * @return true if another such result exists, false otherwise
+             */
+            private boolean findNext() {
+                synchronized (this) {
+                    next = null;
+                    while (getDelegate().hasNext()) {
+                        T n = getDelegate().next();
```

Comment:
It looks like calling:

```
filtered = iter.filter(...);
filtered.hasNext();
filtered.hasNext();
```
would incorrectly consume an element of the filtered results. I think there should be tests to ensure correct behavior of `hasNext` even when called repeatedly.

Maybe `hasNext` needs to check if `next` already is non-null.

Reply:
well spotted